### PR TITLE
Add `_unavailableFromAsync` attribute

### PR DESCRIPTION
**Pull request by etcwilde**
This PR introduces a new underscored attribute `_unavailableFromAsync`, which marks synchronous API as being unavailable from asynchronous contexts. Using an annotated API directly in an asynchronous context will result in a compiler warning message.

https://forums.swift.org/t/discussion-unavailability-from-asynchronous-contexts/53088
rdar://70256865

### CODE REVIEW

**Comment by etcwilde** [comment]
@swift-ci please test

**Comment by etcwilde** [comment]
Windows bot disk is full.

File: include/swift/AST/DiagnosticsSema.def [hunk 1]
```diff
@@ -4699,6 +4699,11 @@ ERROR(actor_isolation_superclass_mismatch,none,
       "%0 class %1 has different actor isolation from %2 superclass %3",
       (ActorIsolation, DeclName, ActorIsolation, DeclName))
 
+ERROR(warn_async_function_must_be_available_from_async,none,
+        "asynchronous function %0 must be available from asynchronous contexts", (DeclName))
+WARNING(warn_async_unavailable_decl,none,
```
Comment (DougGregor) [inline include/swift/AST/DiagnosticsSema.def:4704]: Do we think this should be an error in Swift 6 mode, or will it always be a warning? Or should there be more of a distinction for the user about, e.g., "deprecated" (warning) vs. "unavailable" (error, albeit maybe only in Swift 6)?

File: include/swift/AST/DiagnosticsSema.def [hunk 2]
```diff
@@ -4699,6 +4699,11 @@ ERROR(actor_isolation_superclass_mismatch,none,
       "%0 class %1 has different actor isolation from %2 superclass %3",
       (ActorIsolation, DeclName, ActorIsolation, DeclName))
 
+ERROR(warn_async_function_must_be_available_from_async,none,
```
Comment (DougGregor) [inline include/swift/AST/DiagnosticsSema.def:4702]: Minor nit: it's an error but it's prefixed with `warn_`.

File: lib/Sema/TypeCheckConcurrency.cpp [hunk 3]
```diff
@@ -4282,3 +4282,81 @@ AnyFunctionType *swift::adjustFunctionTypeForConcurrency(
...
+
+/// This is a best-effort to ensure that you aren't directly calling API that is
+/// annotated with an @_unavailableFromAsync attribute.
+/// This is fairly easy to get around. Simply wrapping a call in a closure
+/// will allow calling the API from the synchronous context
+void swift::checkAsyncAvailability(AbstractFunctionDecl &decl) {
+  if (decl.hasAsync() && decl.getAttrs().hasAttribute<UnavailableFromAsyncAttr>())
+    decl.getASTContext().Diags.diagnose(decl.getLoc(),
+        diag::warn_async_function_must_be_available_from_async,
+        decl.getName());
```
Comment (DougGregor) [inline lib/Sema/TypeCheckConcurrency.cpp:4294]: Presumably we need the same kind of check for an unavailable-from-async property, subscript, or initializer, all of which can be `async`. Could you move this to `TypeCheckAttr.cpp`, where we validate the correctness of attributes that are applied to declarations?

File: lib/Sema/TypeCheckAttr.cpp [hunk 4]
```diff
@@ -119,6 +119,7 @@ class AttributeChecker : public AttributeVisitor<AttributeChecker> {
   IGNORED_ATTR(ImplicitSelfCapture)
   IGNORED_ATTR(InheritActorContext)
   IGNORED_ATTR(Isolated)
+  IGNORED_ATTR(UnavailableFromAsync)
```
Comment (DougGregor) [inline lib/Sema/TypeCheckAttr.cpp:122]: ... here's where I think we should do the "you can't put this attribute on an `async` declaration" check.

File: test/ClangImporter/objc_async.swift [hunk 5]
```diff
@@ -43,6 +43,8 @@ func testSlowServer(slowServer: SlowServer) async throws {
   let _: Int = await slowServer.bestName("hello")
   let _: Int = await slowServer.customize("hello")
 
+  slowServer.unavailableMethod() // expected-warning{{'unavailableMethod' is unavailable from asynchronous contexts}}
```
Comment (DougGregor) [inline test/ClangImporter/objc_async.swift:46]: This Just Worked without any code changes? Awesome! Thanks for checking it.

**Review by DougGregor — CHANGES_REQUESTED** [review]

File: lib/Sema/TypeCheckStmt.cpp [hunk 6]
```diff
@@ -2058,6 +2058,7 @@ TypeCheckFunctionBodyRequest::evaluate(Evaluator &evaluator,
   TypeChecker::computeCaptures(AFD);
   if (!AFD->getDeclContext()->isLocalContext()) {
     checkFunctionActorIsolation(AFD);
+    checkAsyncAvailability(*AFD);
```
[the remark this answers is not in the document]
Comment (etcwilde) [reply lib/Sema/TypeCheckStmt.cpp]: The challenge with moving the diagnostic to `diagnoseDeclAvailability` is that we need to be able to ask the `declContext` whether we're directly in an async context. For `AbstractClosureExpr::isAsyncBody`, there needs to be a type assigned to the expression. `diganoseDeclAvailability` is run at a time where we may not have a type or see the expression with a type associated with it.

https://github.com/apple/swift/blob/05fa143a8894bd66c6179abb2928444ae49c04cf/lib/Sema/TypeCheckConstraints.cpp#L400-L418

In `TypeChecker::typeCheckExpression`, we attempt to apply a solution before calling `performSyntacticDiagnosticsForTarget`, which eventually calls into `diagnoseDeclAvailability`. For the simpler test cases, that goes through and the closure expression gets updated with its type. For the more complicated closures like the ones in https://github.com/apple/swift/blob/05fa143a8894bd66c6179abb2928444ae49c04cf/test/Constraints/result_builder.swift, the updated expression doesn't have a type. The constraint system is able to produce a type, but doesn't apply it to the expression until after the children have all type checked, so we don't see the type applied to the decl-context.

https://github.com/apple/swift/blob/cd8da12e0af806642816afefbcd7fd3dbe2c81b3/lib/Sema/CSClosure.cpp#L1393-L1394

There is a type somewhere, though I'm not sure if driving the type from the constraint system, through the availability checking, and into the async checker is the right solution. For the time being, I'm skipping the check and saying that everything is fine.

**Comment by etcwilde** [comment]
@swift-ci please test

**Comment by etcwilde** [comment]
@swift-ci please test

**Comment by etcwilde** [comment]
https://github.com/apple/swift/pull/40341

@swift-ci please test

File: test/ClangImporter/objc_async.swift (hunk 5 above)
Comment (etcwilde) [reply test/ClangImporter/objc_async.swift:48]: Yep, all just workd.

Comment (etcwilde) [reply test/ClangImporter/objc_async.swift:48]: Yes, it all just worked.

**Comment by etcwilde** [comment]
Failure was due to the warning

```
/Users/buildnode/jenkins/workspace/swift-PR-macos/branch-main/swift/test/ClangImporter/objc_async.swift:127:3: error: unexpected warning produced: conformance of 'NonSendableClass' to 'Sendable' is unavailable
  takesSendable(nonSendableClass)     // expected-FIXME-warning{{something about missing conformance}}
  ^
```

CI apparently doesn't pick up branches made to the same repo, so we will have to wait for #40341 to merge before this will pass.

**Comment by etcwilde** [comment]
@swift-ci please test macOS

**Comment by etcwilde** [comment]
@swift-ci please test

**Comment by etcwilde** [comment]
unrelated windows failure.

@swift-ci please test windows

**Comment by etcwilde** [comment]
@swift-ci please test windows

**Comment by etcwilde** [comment]
Requires https://github.com/apple/swift/pull/40368 to pass windows test.

**Comment by etcwilde** [comment]
@swift-ci please test windows

File: include/swift/AST/Attr.def [hunk 7]
```diff
@@ -699,6 +699,12 @@ CONTEXTUAL_SIMPLE_DECL_ATTR(_const, CompileTimeConst,
   ABIStableToAdd | ABIStableToRemove | APIBreakingToAdd | APIStableToRemove,
   126)
 
+SIMPLE_DECL_ATTR(_unavailableFromAsync, UnavailableFromAsync,
+  OnFunc | OnConstructor |
```
Comment (DougGregor) [inline include/swift/AST/Attr.def:703]: I recommend marking this as `UserInaccessible`; that will prevent it from showing up in code completion until we're ready to make it a real feature.

File: lib/Sema/TypeCheckAvailability.cpp [hunk 8]
```diff
@@ -3079,6 +3110,56 @@ bool ExprAvailabilityWalker::diagnoseDeclRefAvailability(
   return false;
 }
 
+/// Diagnose uses of API annotated '@unavailableFromAsync' when used from
+/// asynchronous contexts.
+/// Returns true if a diagnostic was emitted, false otherwise.
+static bool
+diagnoseDeclUnavailableFromAsync(const ValueDecl *D, SourceRange R,
+                                 const Expr *call, const ExportContext &Where) {
+  // FIXME: I don't think this is right, but I don't understand the issue well
```
Comment (DougGregor) [inline lib/Sema/TypeCheckAvailability.cpp:3119]: This is a bit of a depressing comment, but... okay. There's certainly a bug somewhere if we're getting this code before constraint application.

File: test/IDE/complete_decl_attribute.swift [hunk 9]
```diff
@@ -307,6 +311,7 @@ func dummy2() {}
 // KEYWORD_LAST-DAG: Keyword/None:                       transpose[#Declaration Attribute#]; name=transpose
 // KEYWORD_LAST-DAG: Keyword/None:                       noDerivative[#Declaration Attribute#]; name=noDerivative
 // KEYWORD_LAST-DAG: Keyword/None:                       Sendable[#Declaration Attribute#]; name=Sendable
+// KEYWORD_LAST-DAG: Keyword/None:                       _unavailableFromAsync[#Declaration Attribute#]; name=_unavailableFromAsync
```
Comment (DougGregor) [inline test/IDE/complete_decl_attribute.swift:314]: These changes won't be needed if you do the `UserInaccessible` thing.

**Review by DougGregor — APPROVED** [review]
This is looking great!

**Comment by etcwilde** [comment]
@swift-ci please test

File: test/IDE/complete_decl_attribute.swift (hunk 9 above)
Comment (etcwilde) [reply test/IDE/complete_decl_attribute.swift]: Done

File: include/swift/AST/Attr.def (hunk 7 above)
Comment (etcwilde) [reply include/swift/AST/Attr.def]: Done

File: lib/Sema/TypeCheckAvailability.cpp (hunk 8 above)
Comment (etcwilde) [reply lib/Sema/TypeCheckAvailability.cpp:3119]: Yeah, I'm probably going to need some help with someone more familiar with the constraint solver and result builders to get this fully debugged.

**Comment by etcwilde** [comment]
@swift-ci please clean test Linux

**Comment by etcwilde** [comment]
@swift-ci please test Linux

**Pull request merged**